datamol/fp.py

```diff
@@ -1,61 +1,294 @@
 from typing import Union
 from typing import Optional
 
+import warnings
+
 from rdkit import Chem
 from rdkit.Chem import rdMolDescriptors
-from rdkit.Chem import DataStructs
+from rdkit.Chem import rdmolops
+from rdkit.Chem import rdReducedGraphs
+from rdkit.Chem.EState import Fingerprinter as EStateFingerprinter
+from rdkit.Avalon import pyAvalonTools
+
+from rdkit.DataStructs.cDataStructs import ConvertToNumpyArray
+from rdkit.DataStructs.cDataStructs import SparseBitVect
+from rdkit.DataStructs.cDataStructs import UIntSparseIntVect
+from rdkit.DataStructs.cDataStructs import IntSparseIntVect
+from rdkit.DataStructs.cDataStructs import LongSparseIntVect
+from rdkit.DataStructs.cDataStructs import ExplicitBitVect
+from rdkit.DataStructs.cDataStructs import ULongSparseIntVect
 
 import numpy as np
 
 import datamol as dm
 
+_FP_FUNCS = {
+    "maccs": rdMolDescriptors.GetMACCSKeysFingerprint,
+    "ecfp": rdMolDescriptors.GetMorganFingerprintAsBitVect,
+    "topological": rdMolDescriptors.GetHashedTopologicalTorsionFingerprintAsBitVect,
+    "atompair": rdMolDescriptors.GetHashedAtomPairFingerprintAsBitVect,
+    "rdkit": rdmolops.RDKFingerprint,
+    "pattern": rdmolops.PatternFingerprint,
+    "layered": rdmolops.LayeredFingerprint,
+    "erg": rdReducedGraphs.GetErGFingerprint,
+    # NOTE(hadim): bad for pickling?
+    "estate": lambda x, **args: EStateFingerprinter.FingerprintMol(x)[0],
+    "avalon-count": pyAvalonTools.GetAvalonCountFP,
+    # NOTE(hadim): `rdkit-count` to ndarray hangs forever on my machine. Disabling it for now.
+    # "rdkit-count": rdmolops.UnfoldedRDKFingerprintCountBased,
```

Comment:
it's because of the conversion to np array you are doing is very  inefficient  (`from rdkit.DataStructs.cDataStructs import ConvertToNumpyArray`) .  The count based fp is very large (all substructure tested). Ideally  count fp needs to be re-folded again for efficiency as in molfeat, but can do this too:

```
from rdkit.Chem import rdmolops
import numpy as np
fp = rdmolops.UnfoldedRDKFingerprintCountBased(dm.to_mol("CCC"))
fp.GetNonzeroElements() # dict of index, val

tmp = np.zeros(fp.GetLength(), dtype=int)
for n_bit, val in fp.GetNonzeroElements().items():
    tmp[n_bit] = val
fp_out = tmp
```

Reply:
Ok I'll switch to something similar:

```python
tmp = np.zeros(fp.GetLength(), dtype=int)
bit_idx, values = np.array(list(fp.GetNonzeroElements().items())).T
tmp[bit_idx] = values
fp_out = tmp
```

Your snippet actually throw the following error: `ValueError: Maximum allowed dimension exceeded` (as you said because it's unfolded). I wanted `to_fp` to be a direct binding tot he RDKit API so I am not sure what is the best do here:

- remove `rdkit-count`
- perform folding when the fp type is `rdkit-count`
- do nothing and let the error happens: the user can just set `as_array=False`

A preference?

Reply:
On the same example ? GetLength is large, but `tmp = np.zeros(fp.GetLength(), dtype=int)` should not raise an error, unless, it's overflowing np internal array size limit. Re-Folding code is here: 
https://github.com/valence-platform/molfeat/blob/master/molfeat/utils/commons.py#L89

Reply:
> On the same example ?

Yes.

> Re-Folding code is here:

Ok I will add this as an option.

Reply:
> > On the same example ?
> 
> Yes.

I probably have a larger limit then. 

> > Re-Folding code is here:
> Ok I will add this as an option.

Can be disabled too, because I am not sure someone is going to need the count fp for day to day use.

Reply:
> I probably have a larger limit then.

M1 :-)

Reply:
Closing here but check and review the new API please.